src/android/Geolocation.java

```diff
@@ -46,46 +51,109 @@ public class Geolocation extends CordovaPlugin implements OnLocationResultEventL
+        locationContexts.put(id, locationContext);
+
+        if (hasPermission()) {
+            if (checkGMS()) {
+                getLocationGms(locationContext);
+            } else if(checkHMS()) {
+                getLocationHms(locationContext);
             } else {
-                PermissionHelper.requestPermissions(this, id, permissions);
+                popUpGoogleOrHuaweiServicesAreNotPresent();
             }
+        } else {
+            PermissionHelper.requestPermissions(this, id, permissions);
+        }
+    }
 
-        } else if ("addWatch".equals(action)) {
-            int id = args.getString(0).hashCode();
-            LocationContext lc = new LocationContext(id, LocationContext.Type.UPDATE, args, callbackContext, this);
-            locationContexts.put(id, lc);
-
-            if (hasPermission()) {
-                addWatch(lc);
+    /**
+     * Decide if AddWatch will be executed from GMS services (Google Services) or HMS (Huawei Services)
+     *
+     * @param args arguments received from execute
+     * @param callbackContext the callbackContext
+     */
+    private void flowAddWatchGmsOrHms(JSONArray args, CallbackContext callbackContext) throws JSONException {
+        int id = args.getString(0).hashCode();
+        LocationContext lc = new LocationContext(id, LocationContext.Type.UPDATE, args, callbackContext, this,  checkGMS() ? LocationContext.DeviceType.GOOGLE_SERVICES : LocationContext.DeviceType.HUAWEI_SERVICES);
+        locationContexts.put(id, lc);
+
+        if (hasPermission()) {
+            if (checkGMS()) {
+                addWatchGms(lc);
+            } else if(checkHMS()) {
+                addWatchHms(lc);
             } else {
-                PermissionHelper.requestPermissions(this, id, permissions);
+                popUpGoogleOrHuaweiServicesAreNotPresent();
             }
-
-        } else if ("clearWatch".equals(action)) {
-            clearWatch(args, callbackContext);
-
         } else {
-            return false;
+            PermissionHelper.requestPermissions(this, id, permissions);
         }
+    }
 
-        return true;
+    /**
+     * Decide if ClearWatch will be executed from GMS services (Google Services) or HMS (Huawei Services)
+     *
+     * @param args arguments received from execute
+     * @param callbackContext the callbackContext
+     */
+    private void flowClearWatchGmsOrHms(JSONArray args, CallbackContext callbackContext) {
+        String id = args.optString(0);
+
+        if(id != null) {
+            int requestId = id.hashCode();
+            LocationContext lc = locationContexts.get(requestId);
+
+            if(lc == null) {
+                PluginResult result = new PluginResult(PluginResult.Status.ERROR, LocationError.WATCH_ID_NOT_FOUND.toJSON());
```

Comment:
The `result` variable is defined and initialised on both `if` and `else` statements, so it could be defined outside it.